server/src/test/java/org/apache/druid/query/RetryQueryRunnerTest.java

```diff
@@ -0,0 +1,326 @@
+
+    toolChestWarehouse = new QueryToolChestWarehouse()
+    {
+      @Override
+      public <T, QueryType extends Query<T>> QueryToolChest<T, QueryType> getToolChest(final QueryType query)
+      {
+        return conglomerate.findFactory(query).getToolchest();
+      }
+    };
+  }
+
+  @AfterClass
+  public static void tearDownClass() throws IOException
+  {
+    CLOSER.close();
+  }
+
+  @Before
+  public void setup()
+  {
+    segmentGenerator = new SegmentGenerator();
+    httpClient = new TestHttpClient(objectMapper);
+    simpleServerView = new SimpleServerView(toolChestWarehouse, objectMapper, httpClient);
+    cachingClusteredClient = new CachingClusteredClient(
+        toolChestWarehouse,
+        simpleServerView,
+        MapCache.create(0),
+        objectMapper,
+        new ForegroundCachePopulator(objectMapper, new CachePopulatorStats(), 0),
+        new CacheConfig(),
+        new DruidHttpClientConfig(),
+        QueryStackTests.getProcessingConfig(USE_PARALLEL_MERGE_POOL_CONFIGURED),
+        ForkJoinPool.commonPool(),
+        QueryStackTests.DEFAULT_NOOP_SCHEDULER
+    );
+    servers = new ArrayList<>();
+  }
+
+  @After
+  public void tearDown() throws IOException
+  {
+    segmentGenerator.close();
+  }
+
+  private void addServer(DruidServer server, DataSegment dataSegment, QueryableIndex queryableIndex)
+  {
+    servers.add(server);
+    simpleServerView.addServer(server, dataSegment);
+    httpClient.addServerAndRunner(server, new SimpleServerManager(conglomerate, dataSegment.getId(), queryableIndex));
+  }
+
+  @Test
+  public void testNoRetry()
+  {
+    prepareCluster(10);
+    final TimeseriesQuery query = timeseriesQuery(SCHEMA_INFO.getDataInterval());
+    final RetryQueryRunner<Result<TimeseriesResultValue>> queryRunner = createQueryRunner(
+        newRetryQueryRunnerConfig(1, false),
+        query
+    );
+    final Sequence<Result<TimeseriesResultValue>> sequence = queryRunner.run(QueryPlus.wrap(query));
+    final List<Result<TimeseriesResultValue>> queryResult = sequence.toList();
+    Assert.assertEquals(0, queryRunner.getNumTotalRetries());
+    Assert.assertFalse(queryResult.isEmpty());
+    Assert.assertEquals(expectedTimeseriesResult(10), queryResult);
+  }
+
+  @Test
+  public void testRetryForMovedSegment()
+  {
+    prepareCluster(10);
+    final TimeseriesQuery query = timeseriesQuery(SCHEMA_INFO.getDataInterval());
+    final RetryQueryRunner<Result<TimeseriesResultValue>> queryRunner = createQueryRunner(
+        newRetryQueryRunnerConfig(1, true),
+        query
+    );
+    final Sequence<Result<TimeseriesResultValue>> sequence = queryRunner.run(QueryPlus.wrap(query));
+
+    // Let's move a segment
+    dropSegmentFromServerAndAddNewServerForSegment(servers.get(0));
+
+    final List<Result<TimeseriesResultValue>> queryResult = sequence.toList();
+    Assert.assertEquals(1, queryRunner.getNumTotalRetries());
```

Comment:
Is there a guarantee that there will be 1 retry? Why won't the the correct, new server for the segment be selected the first time?

Reply:
Before the last commit, `baseRunner.run()` was executed when `RetryQueryRunner.run()` is called. Since `baseRunner` is the query runner created by `CachingClusteredClient`, `RetryQueryRunner.run()` created the query distribution tree immediately. As a result, I could make the tests more deterministic by emulating moving segments between `sequence = RetryQueryRunner.run()` and `sequence.toList()`.

In the last commit, I changed the code to defer calling `baseRunner.run()` until it has to get executed. This will minimize the chance of segments to move between the time to create query distribution tree and the time to actually send the query to servers. To keep the tests deterministic, I added a new constructor parameter, `runnableAfterFirstAttempt`, which is executed after creating query distribution true for the first attempt. This new parameter must be no-op in production code, but will be used only for testing. I think this parameter will be also useful for integration tests; I can add a new coordinator API to trigger segment balancing and call it after the first attempt using `runnableAfterFirstAttempt`.

Reply:
Oh, I actually found the response context is not shared between subqueries when inlining them. It is shared in UnionQueryRunner though. I think it makes sense to not share the context when inlining subqueries.

Reply:
Is deferring `baseRunner.run()` going to have any unintended consequences? For example, is it going to delay initiation of connections to historicals? I'm not sure, but in general this area of the system is sensitive to when exactly things get called, so it would be good to look into this when changing it.

Reply:
> Is deferring `baseRunner.run()` going to have any unintended consequences? For example, is it going to meaningfully delay initiation of connections to historicals? I'm not sure, but in general this area of the system is sensitive to when exactly things get called, so it would be good to look into this when changing it.

Good question. It does defer initiating connections to query servers if the broker hasn't initiated them yet. AFAIT, there don't seem obvious unintended side effects. [All these query runners running on top of `RetryQueryRunner` are not aware of that the query will be run remotely](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java#L393-L425).